### ISSUE TITLE
Panic with Alternative Struct Definition

### ISSUE THREAD

**Issue by evelritual** [golangci/golangci-lint#3720]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

While linting some older go code we came across an issue with alternative struct definitions using the `=` sign, noted below in the code sample. While I've never seen this way of defining a struct in go (trying to think if I even saw it in version <1.0.0), the compiler doesn't see any issue and afaik works in the same way as a "normal" struct definition without the `=`. The panic seems to be a linter internal to golangci-lint itself, not a specific linter I can define to run, so I did not try a standalone linter but had to mark it off to submit.

Although the linter panic does not reference the line it fails on, once the `=` is removed, the linter works as expected. It seems to me that, while possibly outdated and archaic, it is valid go code and the linter should run normally. It's possible that this code that works but isn't officially supported and we just lucked out on this one, but I'm failing to find any references to this type of definition to see. We'll be removing this way of defining structs, but figured I'd open this in case it'd help someone else in the future. 

This error is reproducible even in a fresh main.go with a main func and type def with an `=`.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.51.2 built from 3e8facb on 2023-02-19T20:54:52Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
run:
  # Timeout for analysis, e.g. 30s, 5m.
  # Default: 1m
  timeout: 3m
  # Ensure running linter doesn't modify go mod
  modules-download-mode: readonly


# This file contains only configs which differ from defaults.
# All possible options can be found here https://github.com/golangci/golangci-lint/blob/master/.golangci.reference.yml
linters-settings:
  exhaustive:
    # Program elements to check for exhaustiveness.
    # Default: [ switch ]
    check:
      - switch
      - map
  nolintlint:
    # Enable to require an explanation of nonzero length after each nolint directive.
    require-explanation: true

  gci:
    custom-order: true
    sections:
      - standard                       # Standard section: captures all standard packages.
      - default                        # Default section: contains all imports that could not be matched to another section type.
      - blank                          # Blank section: contains all blank imports. This section is not present unless explicitly enabled.
      - dot                            # Dot section: contains all dot imports. This section is not present unless explicitly enabled.

  gocritic:
    # Settings passed to gocritic.
    # The settings key is the name of a supported gocritic checker.
    # The list of supported checkers can be find in https://go-critic.github.io/overview.
    settings:
      captLocal:
        # Whether to restrict checker to params only.
        # Default: true
        paramsOnly: false
      underef:
        # Whether to skip (*x).method() calls where x is a pointer receiver.
        # Default: true
        skipRecvDeref: false

  gomnd:
    # List of function patterns to exclude from analysis.
    # Values always ignored: `time.Date`,
    # `strconv.FormatInt`, `strconv.FormatUint`, `strconv.FormatFloat`,
    # `strconv.ParseInt`, `strconv.ParseUint`, `strconv.ParseFloat`.
    # Default: []
    ignored-functions:
      - os.Chmod
      - os.Mkdir
      - os.MkdirAll
      - os.OpenFile
      - os.WriteFile
      - prometheus.ExponentialBuckets
      - prometheus.ExponentialBucketsRange
      - prometheus.LinearBuckets
      - strings.Split
      - time.Ticker
      - make

  gomodguard:
    blocked:
      # List of blocked modules.
      # Default: []
      modules:
        - github.com/golang/protobuf:
            recommendations:
              - google.golang.org/protobuf
            reason: "see https://developers.google.com/protocol-buffers/docs/reference/go/faq#modules"
        - github.com/satori/go.uuid:
            recommendations:
              - github.com/google/uuid
            reason: "satori's package is not maintained"

  govet:
    # Enable all analyzers.
    # Default: false
    enable-all: true
    # Disable analyzers by name.
    # Run `go tool vet help` to see all analyzers.
    # Default: []
    disable:
      - fieldalignment # too strict
      - shadow # usually useless
      - nilness # breaks in >1.18

  nakedret:
    # Make an issue if func has more lines of code than this setting, and it has naked returns.
    # Default: 30
    max-func-lines: 0

  tenv:
    # The option `all` will run against whole test files (`_test.go`) regardless of method/function signatures.
    # Otherwise, only methods that take `*testing.T`, `*testing.B`, and `testing.TB` as arguments are checked.
    # Default: false
    all: true

linters:
  disable-all: true
  enable:
    - asasalint # checks for pass []any as any in variadic func(...any)
    - asciicheck # checks that your code does not contain non-ASCII identifiers
    - bidichk # checks for dangerous unicode character sequences
    - bodyclose # checks whether HTTP response body is closed successfully
    - dupl # tool for code clone detection
    - durationcheck # checks for two durations multiplied together
    - errname # checks that sentinel errors are prefixed with the Err and error types are suffixed with the Error
    - errorlint # finds code that will cause problems with the error wrapping scheme introduced in Go 1.13
    - execinquery # checks query string in Query function which reads your Go src files and warning it finds
    - exhaustive # checks exhaustiveness of enum switch statements
    - exportloopref # checks for pointers to enclosing loop variables
    - forbidigo # forbids identifiers such as fmt.Printf
    - gochecknoinits # checks that no init functions are present in Go code
    - gci # controls golang package import order and makes it always deterministic
    - goconst # finds repeated strings that could be replaced by a constant
    - gocritic # provides diagnostics that check for bugs, performance and style issues
    - goimports # in addition to fixing imports, goimports also formats your code in the same style as gofmt
    - gomnd # detects magic numbers
    - gomodguard # allow and block lists linter for direct Go module dependencies. This is different from depguard where there are different block types for example version constraints and module recommendations
    - gosimple # specializes in simplifying code
    - govet # reports suspicious constructs, such as Printf calls whose arguments do not align with the format string
    - ineffassign # detects when assignments to existing variables are not used
    - lll # reports long lines
    - makezero # finds slice declarations with non-zero initial length
    - nakedret # finds naked returns in functions greater than a specified function length
    - nilerr # finds code that returns nil even if it checks that the error is not nil
    - nilnil # checks that there is no simultaneous return of nil error and an invalid value
    - noctx # finds sending http request without context.Context
    - nonamedreturns # reports all named returns
    - nosprintfhostport # checks for misuse of Sprintf to construct a host with port in a URL
    - predeclared # finds code that shadows one of Go's predeclared identifiers
    - promlinter # checks Prometheus metrics naming via promlint
    - reassign # checks that package variables are not reassigned
    - revive # fast, configurable, extensible, flexible, and beautiful linter for Go, drop-in replacement of golint
    - staticcheck # is a go vet on steroids, applying a ton of static analysis checks
    - stylecheck # is a replacement for golint
    - tenv # detects using os.Setenv instead of t.Setenv since Go1.17
    - tparallel # detects inappropriate usage of t.Parallel() method in your Go test codes
    - typecheck # like the front-end of a Go compiler, parses and type-checks Go code
    - unconvert # removes unnecessary type conversions
    - unparam # reports unused function parameters
    - usestdlibvars # detects the possibility to use variables/constants from the Go standard library
    - unused # checks for unused constants, variables, functions and types
  
    # Might want to enable
    # Ideally should be on but some TODOs are still long-standing
    # - godox # detects FIXME, TODO and other comment keywords
    # Idea is good but unsure how useful it is
    # - gosec # inspects source code for security problems
    # Great for new projects, but hard to pull apart existing code
    # - interfacebloat # checks the number of methods inside an interface
    # Same as above
    # - nestif # reports deeply nested if statements
    # Same as above
    # - gocognit # computes and checks the cognitive complexity of functions
    # Same as above
    # - gocyclo # computes and checks the cyclomatic complexity of functions
    # Same as above
    # - cyclop # checks function and package cyclomatic complexity
    # Same as above
    # - funlen # tool for detection of long functions
    # Broken in >1.18 due to generics
    # - wastedassign # finds wasted assignment statements
    # Same as above
    # - sqlclosecheck # checks that sql.Rows and sql.Stmt are closed
    # Same as above
    # - rowserrcheck # checks whether Err of rows is checked successfully


issues:
  # Maximum count of issues with the same text.
  # Set to 0 to disable.
  # Default: 3
  max-same-issues: 50

  exclude-rules:
    - source: "^//\\s*go:generate\\s"
      linters: [ lll ]
    - source: "(noinspection|TODO)"
      linters: [ godot ]
    - source: "//noinspection"
      linters: [ gocritic ]
    - source: "^\\s+if _, ok := err\\.\\([^.]+\\.InternalError\\); ok {"
      linters: [ errorlint ]
    - path: "metrics/*"
      linters:
        - gomnd
    - path: "_test\\.go"
      linters:
        - bodyclose
        - dupl
        - goconst
        - gomnd
        - lll
        - nilnil
        - noctx

```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.20.1 darwin/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/<redacted>/Library/Caches/go-build"
GOENV="/Users/<redacted>/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/<redacted>/go/pkg/mod"
GONOPROXY="<redacted>"
GONOSUMDB="<redacted>"
GOOS="darwin"
GOPATH="/Users/<redacted>/go"
GOPRIVATE="<redacted>"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.20.1/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.20.1/libexec/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.20.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/<redacted>/go.mod"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/9y/qppn7xy90hvdzdj9zqpy7sxm0000gp/T/go-build137089594=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/<redacted>/go/src/<redacted>/db/spanner /Users/<redacted>/db /Users/<redacted> /Users/<redacted> /Users/<redacted> /Users/<redacted>/go/src /Users/<redacted>/go /Users/<redacted> /Users /] 
INFO [config_reader] Used config file ../../.golangci.yaml 
INFO [lintersdb] Active 43 linters: [asasalint asciicheck bidichk bodyclose dupl durationcheck errname errorlint execinquery exhaustive exportloopref forbidigo gci gochecknoinits goconst gocritic goimports gomnd gomodguard gosimple govet ineffassign lll makezero nakedret nilerr nilnil noctx nonamedreturns nosprintfhostport predeclared promlinter reassign revive staticcheck stylecheck tenv tparallel typecheck unconvert unparam unused usestdlibvars] 
INFO [loader] Go packages loading at mode 575 (name|compiled_files|deps|imports|types_sizes|exports_file|files) took 386.871885ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 4.518874ms 
INFO [linters_context/goanalysis] analyzers took 55.893006266s with top 10 stages: buildir: 29.047272861s, exhaustive: 8.849076566s, inspect: 1.568137642s, fact_deprecated: 1.554561011s, findcall: 1.383850377s, printf: 1.289145462s, ctrlflow: 1.237622641s, fact_purity: 1.140219866s, nilness: 1.087771754s, SA5012: 882.202885ms 
ERRO [runner] Panic: unused: package "spanner" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled case *types.Struct: goroutine 10440 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:24 +0x65
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:102 +0x155
panic({0x100c99600, 0xc02a278850})
        runtime/panic.go:884 +0x213
honnef.co/go/tools/analysis/lint.ExhaustiveTypeSwitch(...)
        honnef.co/go/tools@v0.4.2/analysis/lint/lint.go:234
honnef.co/go/tools/unused.(*graph).embeddedField(0xc00964b618, {0x1010d8d50?, 0xc0035f8f00?}, {0x1010eb158, 0xc00bc7c7d0})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1418 +0x534
honnef.co/go/tools/unused.(*graph).namedType(0xc00964b618, 0x1010d8be8?, {0x1010dcbb0?, 0xc004d1dab8?})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1507 +0x28c
honnef.co/go/tools/unused.(*graph).decl(0xc00964b618, {0x1010dc7f0?, 0xc01dfd72c0}, {0x0, 0x0})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1104 +0xa72
honnef.co/go/tools/unused.(*graph).entry(0xc00964b618)
        honnef.co/go/tools@v0.4.2/unused/unused.go:523 +0xc28
honnef.co/go/tools/unused.run(0xc01242f950)
        honnef.co/go/tools@v0.4.2/unused/unused.go:219 +0x245
github.com/golangci/golangci-lint/pkg/golinters.runUnused(0xc01242f950)
        github.com/golangci/golangci-lint/pkg/golinters/unused.go:61 +0x48
github.com/golangci/golangci-lint/pkg/golinters.NewUnused.func1(0x100d05b40?)
        github.com/golangci/golangci-lint/pkg/golinters/unused.go:31 +0x2f
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc001cf0060)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:188 +0x9df
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:106 +0x1d
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc00214f680, {0x100e69e65, 0x6}, 0xc001b47f48)
        github.com/golangci/golangci-lint/pkg/timeutils/stopwatch.go:111 +0x4a
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0xc0019c64e0?)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:105 +0x85
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc001cf0060)
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb4
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x1eb 
WARN [runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: unused: package "spanner" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled case *types.Struct 
INFO [runner] processing took 3.078µs with stages: nolint: 719ns, skip_dirs: 415ns, max_same_issues: 409ns, cgo: 131ns, autogenerated_exclude: 130ns, max_from_linter: 130ns, path_prettifier: 123ns, exclude-rules: 120ns, source_code: 120ns, skip_files: 116ns, identifier_marker: 113ns, filename_unadjuster: 112ns, sort_results: 63ns, diff: 60ns, exclude: 56ns, max_per_file_from_linter: 54ns, path_shortener: 53ns, uniq_by_line: 52ns, severity-rules: 52ns, path_prefixer: 50ns 
INFO [runner] linters took 8.62811546s with stages: goanalysis_metalinter: 8.628031346s 
ERRO Running error: 1 error occurred:
        * can't run linter goanalysis_metalinter: goanalysis_metalinter: unused: package "spanner" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled case *types.Struct
 
INFO Memory: 92 samples, avg is 570.1MB, max is 992.8MB 
INFO Execution took 9.034257819s
```

</details>

### Code example or link to a public repository

<details>

```go
type sessionID = struct {
	SessionID string
	Placeholder spanner.NullBool
}
```

</details>

**Comment by ldez** [comment]
Hello,

You have checked the following box:

> [x] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

`unused` is related to `staticcheck`, can you give me the output of the run of the standalone binary of [`staticcheck`](https://github.com/dominikh/go-tools)? (golangci-lint v1.51.2 uses `honnef.co/go/tools@v0.4.2`)

https://github.com/golangci/golangci-lint/blob/v1.51.2/go.mod#L115

I think the problem is already fixed inside the latest version of golangci-lint (v1.52.1)

<details>

```go
package main

type sessionID = struct {
	SessionID   string
	Placeholder NullBool
}

type NullBool struct{}

```

```console
$ golangci-lint run                          
main.go:3:6: type `sessionID` is unused (unused)
type sessionID = struct {
     ^

```

</details>

**Comment by evelritual** [comment]
@ldez my mistake, I misread the output. Here's the output, I'll update the original post with this information in a second. Looks to be the same issue between the two static checks defined in v1.52.1(latest?) vs v1.52.2(what we're using). Sorry for the silly question, but am I missing something where v1.52.1 is latest when v.1.52.2 is higher patch version?

Running with staticcheck 0.4.2:
```
panic: internal error: unhandled case *types.Struct

goroutine 2178 [running]:
honnef.co/go/tools/analysis/lint.ExhaustiveTypeSwitch(...)
        honnef.co/go/tools@v0.4.2/analysis/lint/lint.go:234
honnef.co/go/tools/unused.(*graph).embeddedField(0xc002fb5cd0, {0x1604678?, 0xc00118bc40?}, {0x160c878, 0xc0005fd220})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1418 +0x534
honnef.co/go/tools/unused.(*graph).namedType(0xc002fb5cd0, 0x1604538?, {0x16064d0?, 0xc0010cff80?})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1507 +0x28c
honnef.co/go/tools/unused.(*graph).decl(0xc002fb5cd0, {0x1606110?, 0xc00118d240}, {0x0, 0x0})
        honnef.co/go/tools@v0.4.2/unused/unused.go:1104 +0xa72
honnef.co/go/tools/unused.(*graph).entry(0xc002fb5cd0)
        honnef.co/go/tools@v0.4.2/unused/unused.go:523 +0xc28
honnef.co/go/tools/unused.run(0xc0023ff680)
        honnef.co/go/tools@v0.4.2/unused/unused.go:219 +0x245
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc00147f200, {0x16074d8?, 0xc000572c80})
        honnef.co/go/tools@v0.4.2/lintcmd/runner/runner.go:992 +0x76a
honnef.co/go/tools/lintcmd/runner.genericHandle({0x16074d8?, 0xc000572c80}, {0x16074d8?, 0xc00080c000?}, 0x0?, 0xc0003b4b50, 0xc001b2d460)
        honnef.co/go/tools@v0.4.2/lintcmd/runner/runner.go:817 +0x130
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers
        honnef.co/go/tools@v0.4.2/lintcmd/runner/runner.go:1061 +0x70a
```

Running with staticcheck 0.4.0 (what's defined in v1.52.1 go mod):
```
panic: internal error: unhandled case *types.Struct

goroutine 1967 [running]:
honnef.co/go/tools/analysis/lint.ExhaustiveTypeSwitch(...)
        honnef.co/go/tools@v0.4.0/analysis/lint/lint.go:234
honnef.co/go/tools/unused.(*graph).embeddedField(0xc002a1dcd0, {0x16046b8?, 0xc000e025c0?}, {0x160c8b8, 0xc000190a50})
        honnef.co/go/tools@v0.4.0/unused/unused.go:1413 +0x534
honnef.co/go/tools/unused.(*graph).namedType(0xc002a1dcd0, 0x1604578?, {0x1606510?, 0xc000cf2ee8?})
        honnef.co/go/tools@v0.4.0/unused/unused.go:1502 +0x28c
honnef.co/go/tools/unused.(*graph).decl(0xc002a1dcd0, {0x1606150?, 0xc000e00f80}, {0x0, 0x0})
        honnef.co/go/tools@v0.4.0/unused/unused.go:1099 +0xa72
honnef.co/go/tools/unused.(*graph).entry(0xc002a1dcd0)
        honnef.co/go/tools@v0.4.0/unused/unused.go:518 +0xc28
honnef.co/go/tools/unused.run(0xc002099680)
        honnef.co/go/tools@v0.4.0/unused/unused.go:219 +0x245
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000a712f0, {0x1607518?, 0xc001518aa0})
        honnef.co/go/tools@v0.4.0/lintcmd/runner/runner.go:992 +0x76a
honnef.co/go/tools/lintcmd/runner.genericHandle({0x1607518?, 0xc001518aa0}, {0x1607518?, 0xc0014dc320?}, 0x0?, 0xc00037c4d0, 0xc002857e50)
        honnef.co/go/tools@v0.4.0/lintcmd/runner/runner.go:817 +0x130
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers
        honnef.co/go/tools@v0.4.0/lintcmd/runner/runner.go:1061 +0x70a
```

**Comment by ldez** [comment]
You are confusing v1.5**1**.2 and v1.5**2**.1

**Comment by evelritual** [comment]
Ha, you're right, my mistake. Confirmed that v1.52.1 using staticcheck 0.4.3 works, as does standalone. I'll close this issue as it's resolved in latest.
